### PR TITLE
Initial Chunking

### PR DESCRIPTION
**Pull request by nune-tadevosyan**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

- Allows dynamic chunking mechanism for  AED models.
- 
**Collection**: [ASR]

# Changelog

- Add dynamic chunking ability .transcribe()
- Support dynamic chunking with timestamps
- Add unit tests

# Usage
The dynamic chunking feature is automatically enabled when calling `.transcribe()` on a single audio file, or when using `batch_size=1` with multiple audio files that are longer than 40 seconds.
``` python
from nemo.collections.asr.models import EncDecMultiTaskModel
canary_model = EncDecMultiTaskModel.from_pretrained('nvidia/canary-1b-v2')
output=canary_model.transcribe(['<file_path>'],timestamps=True) # or set to False for transcription without timestamps

print("predicted text:", output[0].text)
# word level timestamps
print(output[0].timestamp['word'])

# segment level timestamps
print(output[0].timestamp['segment'])
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 1]
```diff
@@ -561,23 +566,34 @@ def transcribe(
                 )
             trcfg = override_config
             trcfg.timestamps = timestamps
+        # Check if only one audio is provided
+        is_one_audio = isinstance(audio, str) and not (audio.endswith("json") or audio.endswith("jsonl"))
```
Comment (nithinraok) [inline nemo/collections/asr/models/aed_multitask_models.py]: But audio can be list[] with one sample?

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 2]
```diff
@@ -122,6 +124,7 @@ class MultiTaskTranscriptionConfig(TranscribeConfig):
     _internal: Optional[MultiTaskTranscriptionInternalConfig] = field(
         default_factory=lambda: MultiTaskTranscriptionInternalConfig()
     )
+    do_parallel_chunking: bool = False
```
Comment (nithinraok) [inline nemo/collections/asr/models/aed_multitask_models.py]:
```suggestion
    enable_parallel_chunking: bool = False
```

Comment (nithinraok) [reply nemo/collections/asr/models/aed_multitask_models.py]: where is doc string for this parameter?

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 3]
```diff
@@ -976,10 +992,114 @@ def _transcribe_forward(
             batch=batch,
         )
 
+    def _join_y_sequence(self, merged_hypothesis, hypotheses):
```
Comment (nithinraok) [inline nemo/collections/asr/models/aed_multitask_models.py]: move to timestamp utils

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 4]
```diff
@@ -976,10 +992,114 @@ def _transcribe_forward(
...
+        merged_hypothesis.y_sequence = torch.cat([h.y_sequence for h in hypotheses])
+        return merged_hypothesis
+
+    def _join_char_level_timestamps(
```
Comment (nithinraok) [inline nemo/collections/asr/models/aed_multitask_models.py]: move to timestamp utils

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 5]
```diff
@@ -976,10 +992,114 @@ def _transcribe_forward(
...
+
+                char_timestamps.append(upd)
+                j_token += 1
+
+            # 3) make the trim visible to later chunks
+            overall_removed_offset += removed_in_chunk
+
+        return char_timestamps
+
+    def _join_timestamp_and_add_word_and_segment_level_timestamps(
```
Comment (nithinraok) [inline nemo/collections/asr/models/aed_multitask_models.py]: move to timestamp utils

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 6]
```diff
@@ -997,14 +1118,44 @@ def _transcribe_output_processing(self, outputs, trcfg: MultiTaskTranscriptionCo
...
+                    ],  # only approximately 60% of the tokens are non blank
+                    delay=delay,
+                    model=self,
+                    max_steps_per_timestep=1,
+                    min_lcs_length=1,
+                    parallel_chunking=True,
+                )
+                merged_tokens += hypotheses[i].y_sequence.tolist()[int(delay * 0.6) :]
+
+            final_text = self.tokenizer.ids_to_text(merged_tokens)
```
Comment (nithinraok) [inline nemo/collections/asr/models/aed_multitask_models.py]: if necessary create a new file -> chunking_utils.py and move these and other above timestamps merging algo there. Need to keep this model file as clean as possible

File: nemo/collections/asr/parts/mixins/transcription.py [hunk 7]
```diff
@@ -369,13 +372,46 @@ def transcribe_generator(self, audio, override_config: Optional[TranscribeConfig
                     model_outputs = self._transcribe_forward(test_batch, transcribe_cfg)
                     processed_outputs = self._transcribe_output_processing(model_outputs, transcribe_cfg)
 
-                    # clear up memory
-                    del test_batch, model_outputs
-
-                    # Yield results if generator
-                    yield processed_outputs
+                    if is_parallel_chunking:
```
Comment (nithinraok) [inline nemo/collections/asr/parts/mixins/transcription.py]: This file shouldn;t be edited for this task. These needs to be done per model basis.

**Review by nithinraok — CHANGES_REQUESTED** [review]

File: nemo/collections/asr/models/aed_multitask_models.py (hunk 4 above)
Comment (nune-tadevosyan) [reply nemo/collections/asr/models/aed_multitask_models.py]: These are joining timestamps but are actually used only for chunking, so moved to chunking_utils

File: nemo/collections/asr/models/aed_multitask_models.py (hunk 5 above)
Comment (nune-tadevosyan) [reply nemo/collections/asr/models/aed_multitask_models.py]: These are joining timestamps but are actually used only for chunking, so moved to chunking_utils

File: nemo/collections/asr/data/audio_to_text_lhotse_prompted.py [hunk 8]
```diff
@@ -61,22 +62,43 @@ class PromptedAudioToTextLhotseDataset(torch.utils.data.Dataset):
...
 
+        # Will work if batch_size is set to 1.
+        if self.enable_parallel_chunking:
+            # If dynamic chunking is enabled, split each audio sample into chunks.
+            new_audio = []
+            new_audio_lens = []
+            for i in range(audio.shape[0]):
+                waveform = audio[i, : audio_lens[i]]
+                # Split the waveform into chunks and get their lengths.
+                chunks, chunk_lens = self._chunk_waveform(waveform)
```
Comment (nithinraok) [inline nemo/collections/asr/data/audio_to_text_lhotse_prompted.py:95]: Shouldn;t be there an option for overlap control here?

File: nemo/collections/asr/data/audio_to_text_lhotse_prompted.py [hunk 9]
```diff
@@ -61,22 +62,43 @@ class PromptedAudioToTextLhotseDataset(torch.utils.data.Dataset):
...
 
     def __getitem__(self, cuts: CutSet) -> PromptedAudioToTextMiniBatch:
         audio, audio_lens, cuts = self.load_audio(cuts)
 
+        # Will work if batch_size is set to 1.
+        if self.enable_parallel_chunking:
+            # If dynamic chunking is enabled, split each audio sample into chunks.
+            new_audio = []
+            new_audio_lens = []
+            for i in range(audio.shape[0]):
```
Comment (nithinraok) [inline nemo/collections/asr/data/audio_to_text_lhotse_prompted.py:92]: I am guessing you are doing this for future support with bs>1?

File: nemo/collections/asr/data/audio_to_text_lhotse_prompted.py [hunk 10]
```diff
@@ -61,22 +62,43 @@ class PromptedAudioToTextLhotseDataset(torch.utils.data.Dataset):
     Tokenized utterances will be extended with special prompt tokens according to ``prompt_format_fn`` logic.
     We support cuts with multiple supervision segments -- their tokenized texts will be concatenated before we add the prompt tokens.
     This is useful, for example, in code-switched scenarios where each segment is spoken in a different language.
+
+    Chunking:
+    If `enable_parallel_chunking` is True, each audio sample is split into optimally sized chunks
```
Comment (nithinraok) [inline nemo/collections/asr/data/audio_to_text_lhotse_prompted.py]: sorry I think its better to keep it just as 
```suggestion
    If `enable_chunking` is True, each audio sample is split into optimally sized chunks
```

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 11]
```diff
@@ -561,23 +571,39 @@ def transcribe(
...
             trcfg = override_config
             trcfg.timestamps = timestamps
-
-        return super().transcribe(audio=audio, override_config=trcfg)
+        # Check if only one audio is provided with string
+        is_one_audio = isinstance(audio, str) and not (audio.endswith("json") or audio.endswith("jsonl"))
+        # Check if it is provided as a list of strings
+        is_one_audio = is_one_audio or (isinstance(audio, list) and len(audio) == 1)
+        # Check if batch_size is one
+        trcfg.enable_parallel_chunking = is_one_audio or (override_config.batch_size == 1)
```
Comment (nithinraok) [inline nemo/collections/asr/models/aed_multitask_models.py]: Shouldn;t you check all these only when `enable_chunking=True` ? Else if user doesn;t want chunking then this shouldn;t run. Will be useful for parakeet like models

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 12]
```diff
@@ -997,14 +1026,18 @@ def _transcribe_output_processing(self, outputs, trcfg: MultiTaskTranscriptionCo
         decoder_input_ids = outputs.pop('decoder_input_ids')
         batch = outputs.pop('batch')
 
-        del log_probs, encoded_len
-
+        del log_probs
+        num_chunks = enc_states.shape[0]
+        # Repear decoder_input_ids to match number of chunks
+        if trcfg.enable_parallel_chunking and num_chunks > decoder_input_ids.shape[0]:
+            decoder_input_ids = decoder_input_ids.repeat(num_chunks, 1)
```
Comment (nithinraok) [inline nemo/collections/asr/models/aed_multitask_models.py:1036]: why is this required?

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 13]
```diff
@@ -1014,18 +1047,36 @@ def _transcribe_output_processing(self, outputs, trcfg: MultiTaskTranscriptionCo
                 batch_size=len(batch.audio),
                 external_ctc_model=self.timestamps_asr_model,
                 main_model_predictions=hypotheses,
+                timestamp_type='char' if merge_to_be_done else ['word', 'segment'],
```
Comment (nithinraok) [inline nemo/collections/asr/models/aed_multitask_models.py:1053]: currently overlapping LCS through token level?

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 14]
```diff
@@ -1044,6 +1096,8 @@ def _setup_transcribe_dataloader(self, config: Dict) -> 'torch.utils.data.DataLo
             # when using a list of audio files instead of a manifest (added from TranscrptionMixin)
             manifest_filepath = os.path.join(config['temp_dir'], 'manifest.json')
             batch_size = min(config['batch_size'], len(config['paths2audio_files']))
+        # check this part!
+        enable_parallel_chunking = batch_size == 1  # <-- enables chunking for batch size 1
```
Comment (nithinraok) [inline nemo/collections/asr/models/aed_multitask_models.py]: should have an option through `.transcribe(enable_chunking=False)` to disable chunking even when bs=1

File: nemo/collections/asr/models/aed_multitask_models.py (hunk 11 above)
Comment (nune-tadevosyan) [reply nemo/collections/asr/models/aed_multitask_models.py]: This is the point where it gets decided automatically if `enable_chunking`  will work, and it is done based on the batch_size.
We do not want to leave it up to user as for audio files that are short there won't be any difference, but for longer ones results will always be better if we processed it chunk by chunk.
If we allowed user to force it false, then performance will be harmed.

File: nemo/collections/asr/models/aed_multitask_models.py (hunk 12 above)
Comment (nune-tadevosyan) [reply nemo/collections/asr/models/aed_multitask_models.py:1036]: Every entry in the batch of chunks should have its own decoder_input_ids.
For the decoder to work flawlessly and to avoid Error from shape mismatch

File: nemo/collections/asr/models/aed_multitask_models.py (hunk 11 above)
Comment (nithinraok) [reply nemo/collections/asr/models/aed_multitask_models.py]: By default its true, so users who tinker with it should know the consequences for a model

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 15]
```diff
@@ -113,6 +116,10 @@ class MultiTaskTranscriptionInternalConfig(InternalTranscribeConfig):
 class MultiTaskTranscriptionConfig(TranscribeConfig):
     """
     Configuration for Multi Task Transcription
+
+    enable_chunking: bool = False
```
Comment (nithinraok) [inline nemo/collections/asr/models/aed_multitask_models.py]:
```suggestion
    enable_chunking: bool = True
```

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 16]
```diff
@@ -562,22 +571,43 @@ def transcribe(
...
 
     def _setup_dataloader_from_config(self, config: Optional[Dict]):
+
         assert config.get("use_lhotse", False), (
             "Multi-task model only supports dataloading with Lhotse. "
             "Please set config.{train,validation,test}_ds.use_lhotse=True"
         )
         global_rank = config.get("global_rank", self.global_rank)
         world_size = config.get("world_size", self.world_size)
+        enable_chunking = config.get("enable_chunking", False)
```
Comment (nithinraok) [inline nemo/collections/asr/models/aed_multitask_models.py:595]: Set this to true for aed models by default. @monica-sekoyan can you add this to config for the release model.

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 17]
```diff
@@ -1128,6 +1186,8 @@ def _may_be_make_dict_and_fix_paths(self, json_items, manifest_path, trcfg: Mult
                 if k not in entry:
                     # last-chance fallback injecting legacy Canary defaults if none were provided.
                     entry[k] = default_turn.get(k, dv)
+                if k == "timestamp" and self.timestamps_asr_model is not None:
+                    entry[k] = 'notimestamp'
```
Comment (nithinraok) [inline nemo/collections/asr/models/aed_multitask_models.py]: also handle for other `notimestamp` cases like `no` , `False`

File: nemo/collections/asr/parts/utils/aligner_utils.py [hunk 18]
```diff
@@ -111,6 +113,9 @@ def restore_token_case(word: str, word_tokens: List[str]) -> List[str]:
     while "__" in word:
         word = word.replace("__", "_")
 
+    #    while " " in word:
+    #        word = word.replace(" ", "")
+
```
Comment (nithinraok) [inline nemo/collections/asr/parts/utils/aligner_utils.py]:
```suggestion

```

**Review by nithinraok — COMMENTED** [review]
Add unit tests and also one CI test with timestamps and chunking enabled

File: nemo/collections/asr/models/aed_multitask_models.py (hunk 16 above)
Comment (monica-sekoyan) [reply nemo/collections/asr/models/aed_multitask_models.py:595]: as far as I get, this is for transcription config and not for model config, so the model config should not be changed

File: nemo/collections/asr/models/aed_multitask_models.py (hunk 17 above)
Comment (monica-sekoyan) [reply nemo/collections/asr/models/aed_multitask_models.py]: this is corrected in [#14121](https://github.com/NVIDIA/NeMo/pull/14121)

File: nemo/collections/asr/models/aed_multitask_models.py (hunk 16 above)
Comment (nithinraok) [reply nemo/collections/asr/models/aed_multitask_models.py:595]: Its part of `_setup_dataloader_from_config` which is used for setting up during training as well. If default is set to true then this could be an issue

**Review by nithinraok — DISMISSED** [review]

**Review by nithinraok — APPROVED** [review]
